src/Tuura/Plato/Translation.hs

```diff
@@ -76,8 +81,24 @@ validateInterface signs circuit
   where
     unused       = filter ((==Unused) . interface circuit) signs
 
-cartesianProduct :: NonEmpty.NonEmpty [a] -> [[a]]
-cartesianProduct l = sequence (NonEmpty.toList l)
+cartesianProduct :: Ord a => NonEmpty.NonEmpty [a] -> [[a]]
+cartesianProduct l = removeSupersets sortByLength
+  where
+    sequenced    = sequence (NonEmpty.toList l)
+    removeDupes  = map nub sequenced
+    removeNull   = filter (not . null) removeDupes
+    sortAllLists = map sort removeNull
+    sortByLength = sortBy (comparing length) sortAllLists
+
+removeSupersets :: Eq a => [[a]] -> [[a]]
+removeSupersets s = filter (not . null) result
+  where
+    prev n        = take n s
+    check current = any (`isSubsequenceOf` current)
+    result        = map (\x -> if check (s!!x) (prev x)
+                        then []
+                        else s!!x
+                 )[0..(length s) - 1]
```

Comment:
This would be much nicer with list comprehensions:
```haskell
result = [ x | (x:xs) <- tails sortByLength, not (check x xs) ]
```
I guess you'll just need to reverse the sorting order so longer lists come first.